frontend/src/utils/PrivateRoute.tsx

```diff
@@ -1,52 +1,68 @@
+  role?: string | string[];
+  claim?: string | string[];
+}
+interface ComponentRoute extends BaseProtectedAppRoute {
+  customComponent: React.ComponentType<any>;
   layout: React.ComponentType<any>;
-  role?: string | Array<string>;
-  claim?: string | Array<string>;
-  componentProps?: any;
+}
+
+interface RenderRoute extends BaseProtectedAppRoute {
+  customRender: (props: RouteComponentProps<any>) => React.ReactNode;
+}
+
+type IPrivateRouteProps = ComponentRoute | RenderRoute;
+
+function isRenderRoute(route: RenderRoute | ComponentRoute): route is RenderRoute {
+  return (route as RenderRoute).customRender !== undefined;
 }
 
 /**
  * A PrivateRoute only allows a user who is authenticated and has the appropriate role(s) or claim(s).
  * @param props - Properties to pass { component, role, claim }
  */
-const PrivateRoute = (props: IPrivateRouteProps) => {
+const PrivateRoute: React.FC<IPrivateRouteProps> = props => {
   const location = useLocation();
   const keycloak = useKeycloakWrapper();
-  let { component: Component, layout: Layout, ...rest } = props;
-  return (
-    <Route
-      {...rest}
-      render={routeProps => {
-        if (!!keycloak.obj?.authenticated) {
-          if (
-            (!rest.role && !rest.claim) ||
-            keycloak.hasRole(rest.role) ||
-            keycloak.hasClaim(rest.claim)
-          ) {
-            return (
+  let { ...rest } = props;
+
+  if (!!keycloak.obj?.authenticated) {
+    if (
+      (!rest.role && !rest.claim) ||
+      keycloak.hasRole(rest.role) ||
+      keycloak.hasClaim(rest.claim)
+    ) {
+      if (isRenderRoute(props)) {
+        return <Route {...rest} render={props.customRender} />;
+      } else {
+        const { customComponent: Component, layout: Layout } = props;
+        return (
+          <Route
+            {...rest}
+            render={props => (
               <Layout>
-                <Component {...routeProps} {...rest.componentProps} />
+                <Component {...props} />
               </Layout>
-            );
-          } else {
-            return (
-              <Redirect to={{ pathname: '/forbidden', state: { referer: routeProps.location } }} />
-            );
-          }
-        } else {
-          if (routeProps.location.pathname !== '/login') {
-            const redirectTo = encodeURI(`${location.pathname}${location.search}`);
-            return <Redirect to={`/login?redirect=${redirectTo}`} />;
-          }
-        }
-      }}
-    />
-  );
+            )}
+          />
+        );
+      }
+    } else {
+      return <Redirect to={{ pathname: '/forbidden', state: { referer: location } }} />;
+    }
+  } else {
+    if (location.pathname !== '/login') {
+      const redirectTo = encodeURI(`${location.pathname}${location.search}`);
+      return <Redirect to={`/login?redirect=${redirectTo}`} />;
+    } else {
+      console.error('Navigation: location was expected to be login');
+      return <></>;
```

Comment:
maybe the error page instead?

Reply:
We could revisit it. As it is it maintains the same functionality as before. i.e. not showing anything